rllib/algorithms/ppo/ppo.py

```diff
@@ -416,90 +427,125 @@ def training_step(self):
                     worker_set=self.workers,
                     max_env_steps=self.config.total_train_batch_size,
                     sample_timeout_s=self.config.sample_timeout_s,
                     _uses_new_env_runners=self.config.uses_new_env_runners,
+                    _return_metrics=True,
                 )
             # Return early if all our workers failed.
             if not episodes:
                 return {}
-            self._counters[NUM_AGENT_STEPS_SAMPLED] += sum(
-                e.agent_steps() for e in episodes
-            )
-            self._counters[NUM_ENV_STEPS_SAMPLED] += sum(
-                e.env_steps() for e in episodes
-            )
 
-        # Perform a train step on the collected batch.
-        train_results = self.learner_group.update_from_episodes(
-            episodes=episodes,
-            minibatch_size=(
-                self.config.mini_batch_size_per_learner
-                or self.config.sgd_minibatch_size
-            ),
-            num_iters=self.config.num_sgd_iter,
-        )
+            # Reduce EnvRunner metrics over the n EnvRunners.
+            self.metrics.log_n_dicts(env_runner_metrics, key=ENV_RUNNER_RESULTS)
+            # Log lifetime counts for env- and agent steps.
+            self.metrics.log_dict(
+                {
+                    NUM_AGENT_STEPS_SAMPLED_LIFETIME: self.metrics.peek(
+                        ENV_RUNNER_RESULTS, NUM_AGENT_STEPS_SAMPLED
+                    ),
+                    NUM_ENV_STEPS_SAMPLED_LIFETIME: self.metrics.peek(
+                        ENV_RUNNER_RESULTS, NUM_ENV_STEPS_SAMPLED
+                    ),
+                    NUM_EPISODES_LIFETIME: self.metrics.peek(
+                        ENV_RUNNER_RESULTS, NUM_EPISODES
+                    ),
+                },
+                reduce="sum",
+            )
 
-        # The train results's loss keys are pids to their loss values. But we also
-        # return a total_loss key at the same level as the pid keys. So we need to
-        # subtract that to get the total set of pids to update.
-        # TODO (Kourosh): We should also not be using train_results as a message
-        #  passing medium to infer which policies to update. We could use
-        #  policies_to_train variable that is given by the user to infer this.
-        policies_to_update = set(train_results.keys()) - {ALL_MODULES}
+        # Perform a learner update step on the collected episodes.
+        with self.metrics.log_time((TIMERS, LEARNER_UPDATE_TIMER)):
+            learner_results = self.learner_group.update_from_episodes(
+                episodes=episodes,
+                minibatch_size=(
+                    self.config.mini_batch_size_per_learner
+                    or self.config.sgd_minibatch_size
+                ),
+                num_iters=self.config.num_sgd_iter,
+            )
+            self.metrics.log_dict(learner_results, key=LEARNER_RESULTS)
```

Comment:
Correct: `learner_results` is coming from the LearnerGroup, which still uses the old `reduce_fn` style reduction to return a single result dict (from all the n workers' result dicts). We have to do a follow up PR, in which I would like to:
* Return n result dicts from LearnerGroup (I don't think LearnerGroup should have its own metrics logger object), instead of just one.
* Then, yes, use `log_n_dicts` inside Algorithm to further reduce over the n Learner workers. Then the Learner metrics get reduced in exactly the same way as the m EnvRunner metrics.